### PR TITLE
server: do not block catalog migration on inconsistent metadata

### PR DESCRIPTION
**Pull request by codingkarthik**
Earlier, we used to terminate the graphql-engine process during the startup if there were any inconsistent metadata, this only happened when there were some catalog migration(s). In the other case, when there are no catalog migrations there to be applied, we don't terminate the process, this PR replicates the same behaviour for when there are catalog migration(s) to be applied.

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System
- [ ] Tests
- [ ] Other (list it)

### Steps to test and verify

To manually test this,

1. Setup a local GraphQL server and add it as a remote schema in the graphql-engine.
2. Shut down the graphql-engine create a dummy migration and increment the `catalog_version.txt`.
3. Shutdown the local GraphQL server and then start the graphql-engine.
4. The graphql-engine should have been started up successfully.

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [x] No
- [ ] Yes
  - [ ] Updated docs with SQL for downgrading the catalog 

#### Metadata

Does this PR add a new Metadata feature?
- [x] No
- [ ] Yes
  - Does `run_sql` auto manages the new metadata through schema diffing?
    - [ ] Yes
    - [ ] Not required
  - Does `run_sql` auto manages the definitions of metadata on renaming?
    - [ ] Yes
    - [ ] Not required
  - Does `export_metadata`/`replace_metadata` supports the new metadata added?
    - [ ] Yes
    - [ ] Not required

#### GraphQL
- [ ] No new GraphQL schema is generated
- [ ] New GraphQL schema is being generated:
   - [ ] New types and typenames are correlated
   
   

#### Breaking changes

- [ ] No Breaking changes
- [ ] There are breaking changes:

  1. Metadata API

     Existing `query` types:
     - [ ] Modify `args` payload which is not backward compatible
     - [ ] Behavioural change of the API
     - [ ] Change in response `JSON` schema
     - [ ] Change in error code
     

  2. GraphQL API

     Schema Generation:
     
     
     - [ ] Change in any `NamedType`
     - [ ] Change in table field names
     

     Schema Resolve:-
     
     - [ ] Change in treatment of `null` value for any input fields 
     

  3. Logging

     - [ ] Log `JSON` schema has changed
     - [ ] Log `type` names have changed

### CODE REVIEW

File: server/src-lib/Hasura/Server/Migrate.hs [hunk 1]
```diff
@@ -164,7 +164,9 @@ migrateCatalog env migrationTime = do
     buildCacheAndRecreateSystemMetadata :: m (RebuildableSchemaCache m)
     buildCacheAndRecreateSystemMetadata = do
       schemaCache <- buildRebuildableSchemaCache env
-      view _2 <$> runCacheRWT schemaCache recreateSystemMetadata
+      case scInconsistentObjs (lastBuiltSchemaCache schemaCache) of
+        [] -> view _2 <$> runCacheRWT schemaCache recreateSystemMetadata
+        _  -> pure schemaCache
```
Comment (codingkarthik) [inline server/src-lib/Hasura/Server/Migrate.hs:169]: I'm not sure if this is the right solution. I tried to replicate the behaviour of when `previousVersion == latestCatalogVersionString` in the `migrateFrom` function in the `migrateCatalog` function. 

Calling the `runCacheRWT` function with a `schemaCache` containing inconsistent metadata throws an error, I don't understand the `runCacheRWT` function well enough to know why that happens.

@lexi-lambda Please do let me know if there is a better way to solve this issue 😄

Comment (lexi-lambda) [reply server/src-lib/Hasura/Server/Migrate.hs:169]: This is definitely *not* the right solution: we can’t just skip updating the system metadata if there are inconsistent objects! Then the system metadata might be wrong/out of date, and it wouldn’t be fixed until the next migration.

The reason an exception is raised is that `recreateSystemMetadata` calls `buildSchemaCacheStrict`, which raises an error on inconsistent metadata. Fortunately, this has a simple solution: call `buildSchemaCache` instead, which does not raise an error on inconsistent metadata.

**Review by lexi-lambda — CHANGES_REQUESTED** [review]

File: server/src-lib/Hasura/Server/Migrate.hs (hunk 1 above)
Comment (codingkarthik) [reply server/src-lib/Hasura/Server/Migrate.hs]: Oh, I didn't know that's what `buildSchemaCache` is used for 😅 . I have fixed this. Thanks!

**Review by 0x777 — APPROVED** [review]

File: server/src-lib/Hasura/Server/Migrate.hs (hunk 1 above)
Comment (lexi-lambda) [reply server/src-lib/Hasura/Server/Migrate.hs]: No worries! No harm no foul, and the updated patch looks good to me.

**Review by lexi-lambda — APPROVED** [review]

**Pull request closed without merging**